File: kweb3-multicall/src/main/java/dev/klepto/kweb3/multicall/MulticallContract.java
package dev.klepto.kweb3.multicall;

import dev.klepto.kweb3.core.Web3Result;
import dev.klepto.kweb3.core.contract.Web3Contract;
import dev.klepto.kweb3.core.type.EthAddress;
import dev.klepto.kweb3.core.type.EthArray;
import dev.klepto.kweb3.core.type.EthBytes;
import dev.klepto.kweb3.core.type.EthUint;

/**
 * Declares a common multicall contract signature to encapsulate most common multicall contract implementations.
 *
 * @author <a href="http://github.com/klepto">Augustinas R.</a>
 */
public interface MulticallContract extends Web3Contract {

    /**
     * Executes a multicall request with given parameters. As there is no-way to enforce method implementation for
     * contract proxy interfaces, this method is meant to be overridden using `default` keyword with underlying
     * multicall implementation.
     *
     * @param gasLimit  the gas limit of single call execution
     * @param sizeLimit the size limit of single call execution
     * @param addresses the list of contract addresses
     * @param data      the list of call-data to be sent to given addresses
     * @return an ordered array containing resulting bytes (possibly empty, if call failed) of each call
     */
    Web3Result<EthArray<EthBytes>> execute(
            EthUint gasLimit,
            EthUint sizeLimit,
            EthArray<EthAddress> addresses,
            EthArray<EthBytes> data
    );

    /**
     * Creates a new {@link MulticallBuilder} for this multicall smart contract implementation.
     *
     * @return a new instance of multicall builder
     */
    default MulticallBuilder builder() {
        return new MulticallBuilder(this);
    }

}

File: kweb3-kotlin/src/main/kotlin/dev/klepto/kweb3/kotlin/multicall/MulticallContract.kt
package dev.klepto.kweb3.kotlin.multicall

import dev.klepto.kweb3.core.contract.Web3Contract
import dev.klepto.kweb3.core.contract.type.EthStructContainer
import dev.klepto.kweb3.core.ethereum.type.EthValue
import dev.klepto.kweb3.core.ethereum.type.primitive.EthAddress
import dev.klepto.kweb3.core.ethereum.type.primitive.EthBytes
import dev.klepto.kweb3.kotlin.CoroutineContractExecutor

/**
 * Executes multiple calls to the blockchain in a single transaction
 * through a use of *multicall* smart contract.
 *
 * @author <a href="http://github.com/klepto">Augustinas R.</a>
 */
interface MulticallContract : Web3Contract {
    /**
     * Executes multiple calls to the blockchain in a single transaction and
     * returns list of their return data. If [allowFailure] is set to `false`,
     * any failed call will cause this function to fail. If [allowFailure] is
     * set to `true`, failed calls will contain `null` return data in their
     * respective index of the resulting list.
     *
     * @param allowFailure signifies whether to allow individual calls to fail
     * @param calls the list of smart contract calls
     * @return an ordered list of return data from each call, possibly `null`
     *     for failed calls
     */
    suspend fun execute(
        allowFailure: Boolean,
        calls: List<Call>,
    ): List<EthBytes?>

    /**
     * A single smart contract call to be executed in a *multicall*
     * transaction.
     *
     * @param address the address of the smart contract
     * @param data the encoded call data represented as ethereum bytes
     */
    data class Call(
        val address: EthAddress,
        val data: EthBytes,
    ) : EthStructContainer

    companion object {
        /**
         * Gets the underlying [CoroutineContractExecutor] used to execute this
         * *multicall* transactions.
         */
        fun MulticallContract.contractExecutor(): CoroutineContractExecutor {
            val contractExecutor = client.contractExecutor
            require(contractExecutor is CoroutineContractExecutor) {
                "${this::class.simpleName} requires a ${CoroutineContractExecutor::class.simpleName}."
            }
            return contractExecutor
        }

        /**
         * Creates a new [MulticallBuilder2] for constructing a *multicall* request
         * using this *multicall* executor.
         *
         * @return a new multicall builder
         */
        inline fun <reified T : EthValue> MulticallContract.builder(): MulticallBuilder<T> {
            return MulticallBuilder(this)
        }
    }
}
